tutorials/native-debugging.md

```diff
@@ -71,4 +71,45 @@ settings append target.exec-search-paths <PATH>/objdir-android-opt/mozglue/build
 ```lldb
 b <file>.cpp:<line number>
 ```
-5. Once your breakpoints have been set, click the continue execution button to move beyond the `ElfLoader` breakpoint and your newly set native breakpoints should be hit. Debug as usual.
\ No newline at end of file
+5. Once your breakpoints have been set, click the continue execution button to move beyond the `ElfLoader` breakpoint and your newly set native breakpoints should be hit. Debug as usual.
+
+## Attaching debuggers to content and other child processes
+
+Internally, GeckoView has a multi-process architecture.  The main Gecko process lives in the main Android process, but content rendering and some other functions live in child processes.  This balances load, ensures certain critical security properties, and allows GeckoView to recover if content processes become unresponsive or crash.  However, it's generally delicate to debug child processes because they come and go.
+
+The general approach is to make the Java code in the child process that we want to debug wait for a Java debugger at startup, and then to connect such a Java debugger manually from the Android Studio UI.
+
+[Bug 1522318](https://bugzilla.mozilla.org/show_bug.cgi?id=1522318) tracks making this debug process more developer-friendly.
+
+### Making a child process wait for a Java debugger
+
+Android itself provides [waitForDebugger](https://developer.android.com/reference/android/os/Debug.html#waitForDebugger()).  We just need to invoke it in the correct process at the correct time.
+
+A good place to do that is in in [`org.mozilla.gecko.process.GeckoServiceChildProcess.onCreate()`](https://searchfox.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/gecko/process/GeckoServiceChildProcess.java#43).  For example, make `onCreate()` look like:
```

Comment:
I think we should have a flag for this so that we don't have to change the code to enable it (especially for embedders who hopefully don't build geckoview themselves). I added a comment to Bug 1522318.